### PR TITLE
use methods to find companion class

### PR DESCRIPTION
**Pull request by DarkDimius**

### CODE REVIEW

**Comment by DarkDimius** [comment]
This branch includes change only to `companionClass`. I'll make a separate PR for `companionModule` when I implement it.

**Comment by DarkDimius** [comment]
@odersky please review.

File: src/dotty/tools/dotc/ast/Desugar.scala [hunk 1]
```diff
@@ -361,9 +361,11 @@ object desugar {
         }
         companionDefs(parent, applyMeths ::: unapplyMeth :: defaultGetters)
       }
-      else if (defaultGetters.nonEmpty)
-        companionDefs(anyRef, defaultGetters)
-      else Nil
+      else {
```
Comment (odersky) [inline src/dotty/tools/dotc/ast/Desugar.scala]: Why the change in formatting? I liked the old one better.

File: src/dotty/tools/dotc/ast/Desugar.scala [hunk 2]
```diff
@@ -409,6 +411,8 @@ object desugar {
     flatTree(cdef1 :: companions ::: implicitWrappers)
   }
 
+  val AccessOrSynthetic = AccessFlags | Synthetic
```
Comment (odersky) [inline src/dotty/tools/dotc/ast/Desugar.scala]: Rename to SyntheticAccessor. When you set mods, its not an Or.

File: src/dotty/tools/dotc/ast/Desugar.scala (hunk 1 above)
Comment (DarkDimius) [inline src/dotty/tools/dotc/ast/Desugar.scala]: This code was rewritten, it used to have more complicated logic between the commits. I have no problem changing it back.

File: src/dotty/tools/dotc/core/StdNames.scala [hunk 3]
```diff
@@ -117,13 +117,15 @@ object StdNames {
...
-    val SHADOWED: N                  = "(shadowed)"  // tag to be used until we have proper name kinds
+    val SHADOWED: N                   = "(shadowed)"  // tag to be used until we have proper name kinds
     val SINGLETON_SUFFIX: N           = ".type"
     val SPECIALIZED_SUFFIX: N         = "$sp"
     val SUPER_PREFIX: N               = "super$"
     val WHILE_PREFIX: N               = "while$"
     val DEFAULT_EXCEPTION_NAME: N     = "ex$"
     val INITIALIZER_PREFIX: N         = "initial$"
+    val COMPANION_MODULE_METHOD: N    = "companion$module"
+    val COMPANION_CLASS_METHOD: N     = "compaion$class"
```
Comment (odersky) [inline src/dotty/tools/dotc/core/StdNames.scala]: Typo, should be "companion$class"

File: src/dotty/tools/dotc/ast/Desugar.scala (hunk 2 above)
Comment (DarkDimius) [inline src/dotty/tools/dotc/ast/Desugar.scala]: I'm not setting them, I'm using this value to mask previous mods.

File: src/dotty/tools/dotc/core/SymDenotations.scala [hunk 4]
```diff
@@ -773,11 +773,31 @@ object SymDenotations {
...
+      */
+    final def companionClass(implicit ctx: Context): Symbol = {
+      val companionMethod = info.decls.denotsNamed(nme.COMPANION_CLASS_METHOD, selectPrivate).first
+
+      if (companionMethod.exists)
+        companionMethod.info.resultType.classSymbol
+      else {
+        /*
+        val scalac = companionNamed(effectiveName.toTypeName)
+
```
Comment (odersky) [inline src/dotty/tools/dotc/core/SymDenotations.scala]: Is the commented out code still needed? If not, delete as well as the enclosing {...}

File: src/dotty/tools/dotc/core/Symbols.scala [hunk 5]
```diff
@@ -161,6 +161,15 @@ trait Symbols { this: Context =>
...
+  def synthesizeCompanionMethod(name: TermName, ret: SymDenotation, owner: SymDenotation)(implicit ctx: Context) = {
+    if(owner.exists && ret.exists) ctx.newSymbol(
+      owner = owner.symbol,
+      name = name,
+      flags = Flags.Synthetic | Flags.Private,
+      info = ExprType(ret.typeRef))
+    else NoSymbol
+  }
+
   /** Create a package symbol with associated package class
```
Comment (odersky) [inline src/dotty/tools/dotc/core/Symbols.scala]: - Add a doc comment. 
- Formatting: if (.
- It's more common to pass owner as a symbol
- "ret" is not descriptive.
- In this case, I do not think the named parameters add anything. All arguments make it clear by their names already what parameter they are for.

I'd reformulate as follows:

    def synthesizeCompanionMethod(target: Symbol, owner: Symbol)(implicit ctx: Context) = {
       val name = if (target is Module) nme.COMPANION_MODULE_METHOD else nme.COMPANION_CLASS_METHOD
       ctx.newSymbol(owner, name, Flags.Synthetic | Flags.Private, ExprType(target.typeRef)
    }

File: test/dotc/tests.scala [hunk 6]
```diff
@@ -149,7 +149,7 @@ class tests extends CompilerTest {
 
   @Test def dotc_printing = compileDir(dotcDir + "tools/dotc/printing")
 
-  @Test def dotc_reporting = compileDir(dotcDir + "tools/dotc/reporting", twice)
+  @Test def dotc_reporting = compileDir(dotcDir + "tools/dotc/reporting")
```
Comment (odersky) [inline test/dotc/tests.scala]: Why was the twice removed here?

File: src/dotty/tools/dotc/typer/Namer.scala [hunk 7]
```diff
@@ -395,32 +395,64 @@ class Namer { typer: Typer =>
...
                     rhs = cpy.Template(impl)(body = compimpl.body ++ impl.body))
                 mdef.putAttachment(ExpandedTree, Thicket(vdef :: mcls1 :: Nil))
+                moduleDef(typName) = mcls1
                 cdef.putAttachment(ExpandedTree, Thicket(cls :: crest))
               case _ =>
             }
           case none =>
         }
+      }
+    }
```
Comment (odersky) [inline src/dotty/tools/dotc/typer/Namer.scala]: I think this is a bit on the complex side, in particuar I do not like the repetition of the complicated Thicket pattern matches. Can we experiment with having 4 Maps:

 `classDef`, `moduleDef` as before

     val syntheticCompanionDef: Map[TypeName, TypeDef]  // a map from case classes to synthetic companions
     val coreCaseClassDef: Map[TypeName, List[Tree]] // a map from case classes to the definitions w/o synthetic companions

Using these we could streamline the traversal of ModuleDefs. 

See what I mean?

**Comment by odersky** [comment]
I did not see a place where companion methods are turned into trees. Did I miss something? The concern here is that TreeTypeMap (and with it, subst, changeOwner) work on trees, not scopes. So they will not create new companion links unless the methods are part of the tree,

File: src/dotty/tools/dotc/typer/Namer.scala (hunk 7 above)
Comment (DarkDimius) [inline src/dotty/tools/dotc/typer/Namer.scala]: Should those maps be directly updated by desugar or you'll still use attachments?
If second, you'll still need 2 pattern matches, one over classes, and one over modules to populate maps.
And then you'll need an additional pattern match to update the annotations, so you still have three pattern matches. Am I missing something?

**Comment by DarkDimius** [comment]
Actually due to way how `TreeTypeMap` works now, I'm not synthesizing the tree. It's not obvious what does it mean for companion links to `TreeTypeMap` the tree, which includes a class but doesn't include the companion. In this case we can end up with several classes that have the same companion module, eg in case of
```
case class C[@specialized T](a: T)
```
we'll have companion object which is same for all specialized classes.

I wanted to discuss if this is what we want. Actually the more i think of this, the more I believe that this actually makes sense.

File: src/dotty/tools/dotc/core/pickling/UnPickler.scala [hunk 8]
```diff
@@ -118,6 +118,13 @@ object UnPickler {
         denot.owner.thisType select denot.sourceModule
       else selfInfo
     if (!(denot.flagsUNSAFE is JavaModule)) ensureConstructor(denot.symbol.asClass, decls)
+    if (denot.flagsUNSAFE is Module) {
+      val scalacCompanion = denot.classSymbol.scalacLinkedClass
+      val companionClassMethod = ctx.synthesizeCompanionMethod(nme.COMPANION_CLASS_METHOD, scalacCompanion, denot.classSymbol)
+      if (companionClassMethod.exists)
+        companionClassMethod.entered
+    }
+
```
Comment (odersky) [inline src/dotty/tools/dotc/core/pickling/UnPickler.scala]: It's a pity that we still need scalacLinkedClass here, because it means we cannot get rid of the associated complexity of isCoDefinedWith. I would much prefer if we got the linked class directly from the unpickling info, instead of looking at 
the scope.

**Comment by odersky** [comment]
Overall, I am a bit ambivalent about this PR. It adds a lot of complications. So far the main benefit is that it makes linkedClass insensitive to naming issues. However, I would prefer simply to get naming right instead of inventing another mechanism to patch any faults it might have. And it looks that this (getting naming right) requires altogether less effort than this PR.

There would be a win in my mind if we could at the same time get rid of isCoDefinedWith which is fragile and very dependent on how the compiler is used. But so far the PR does not do that.

**Comment by DarkDimius** [comment]
@odersky, can you please tell which complications you mean in your comment? All this pr does:
 1. Detects companion links using old scheme for Java and Scala2 defined classes.
 2. Adds companion links for dotty-defined classes during appropriate moment that doesn't break anything in typer.


1. relies indeed currently relies on previous implementation, but if there's a simpler way to discover companion links, nothing stops us from using it.
2. doesn't seem to add any complications, but actually revealed several bugs\deviations from what scalac does

**Comment by DarkDimius** [comment]
>  So far the main benefit is that it makes linkedClass insensitive to naming issues. However, I would prefer simply to get naming right instead of inventing another mechanism to patch any faults it might have.

Yes, and it also gives us guarantees that any tool that somehow changes or interacts with TASTY files, or any compiler plugin, becomes easier to implement as there is one less thing that they can break: links between companions cannot be silently broken without intention anymore. I believe that this is an improvement on its own.

**Comment by odersky** [comment]
Regarding the TreeTypeMap issue, the problem is if you duplicate a code like this:

     def f() { class C { ... } object C { ... } ... }

I do not see how the link between the copies of the C's could be maintained.

**Comment by odersky** [comment]
Complications: I just look at (1) lines added, (2) files touched, and the pretty hairy code in Namer. It certainly looks like we cannot contain this to a single point, and have to add complications to severas places. 

Regarding robustness, I am still not sure. I am more of the school of "fail early". That is, if somebody gets naming wrong it's better to fail immediately rather than have another mechanism that works around it. Otherwise put, try to get to a state where every piece of information is encoded in exactly one way. One could argue that companion methods are what _defines_ companionship and names do not matter. But then we need to (1) get rid of references where names do matter as in scalacLinkedClass and (2) make sure that names will never matter in any other way. If we fail to do that, it's probably better to make sure names work right, because that way we do not duplicate info and if we get it wrong we fail early.

But I noted another aspect where the new scheme beats the old: We no longer need to traverse the tree to find companions of inner classes. That's a big win! And furthermore, one could argue that scalac unpickling is the only reason we still need scalaLinkedClass, and hopefully that will go away and be replaced with TASTY. So that tilts the balance back in favor of this PR.

**Comment by DarkDimius** [comment]
@odersky, in your example, are you TreeTypeMap-ing `def f()`? or only `class C`? 
If first one(`def f()`), it should use the same mechanism that works for every method in `class C` that could return `object C`. I've tested and it just works(`withMappedSyms` in `transformDefs` handles this).


If you are duplicating `class C` inside `f` without duplicating `object C` we need to understand what does this mean for those links at all.

I propose to take this discussion offline, but it would help me if you could tell me what do you find complicated in this PR, as I will have time to address this comments and come up with simpler implementation.

The only complicated thing that I see around is `scalacLinkedClass`, which is not added but only maintained. It is only used to retrieve the information from `scalac`, if we had TASTY, it would no longer be needed.

File: src/dotty/tools/dotc/ast/Desugar.scala (hunk 2 above)
Comment (odersky) [inline src/dotty/tools/dotc/ast/Desugar.scala]: Ah, I misread. Consider the comment withdrawn.

**Comment by odersky** [comment]
I meant: duplicate `def f`. On second thought, I now see how this works. So much the better if we do not need the trees, since there's no code generated for them anyway.

**Comment by odersky** [comment]
Overall I am now in favor of the PR. I would try hard to find a better implementation in Namer, however.

**Comment by odersky** [comment]
> I've added the assertion, based on discussion that we had: if `enter` gets a scope as an argument, than
> this is a scope that will eventually become `decls` of this symbol. And this should only happen if this is
> first time the scope of symbol is coputed, ie symbol yet has no future. 

OK, good! Then let's add this to the comment, to make it more understandable.

**Comment by odersky** [comment]
That's all I have.

File: src/dotty/tools/dotc/core/Symbols.scala (hunk 5 above)
Comment (DarkDimius) [inline src/dotty/tools/dotc/core/Symbols.scala]: Defining method name based on flags doesn't work here: in order to not force type you will need `flagsUNSAFE`, and tests suggest that Module could still be unset.
